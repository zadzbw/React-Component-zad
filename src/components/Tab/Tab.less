@import "../../style/variables";
@import "../../style/mixins/index";

@tab-prefix: zad-tab;
@bar-padding: 5px;

.@{tab-prefix} {
  .clearfix();
  overflow: hidden;

  &-bars {
    &-list {
      border-bottom: 1px solid @border-color-base;
      margin-bottom: 16px;
      outline: none;
    }

    &-container {
      position: relative;
      overflow: hidden;
      width: 100%;
      height: 32px;
      margin-bottom: -1px;
    }

    &-wrap {
      overflow: hidden;
    }

    &-bottom {
      display: block;
      position: absolute;
      height: 2px;
      left: 0;
      bottom: 0;
      z-index: 1;
      background-color: @primary-color;
      will-change: transfrom;
      transform-origin: 0 0;
    }
  }

  &-bar {
    display: inline-block;
    height: 100%;
    color: @text-color;
    user-select: none;
    cursor: pointer;
    margin: 0 2px 0 0;
    padding: @bar-padding 16px;
    border: 1px solid @border-color-base;
    border-bottom: 0;
    border-radius: @border-radius-base @border-radius-base 0 0;
    transition: all 0.3s @ease-in-out;
    background: #f9f9f9;
    position: relative;
    box-sizing: border-box;

    &:not(.@{tab-prefix}-bar-active):hover {
      color: tint(@primary-color, 10%);
    }

    &:not(.@{tab-prefix}-bar-active):active {
      color: shade(@primary-color, 10%);
    }

    &-active {
      color: @primary-color;
      background-color: @body-background;
      transform: translateZ(0);
      padding-bottom: @bar-padding + 1;
    }
  }

  &-content {
    display: flex;
    will-change: margin-left;
    flex-direction: row;
    width: 100%;
  }

  &-item {
    display: inline-block;
    flex-shrink: 0; // 不会缩小
    width: 100%;
    transition: opacity 0.3s @ease-in-out;

    &-active {
      opacity: 1;
    }

    &-inactive {
      height: 0;
      opacity: 0;
    }
  }
}
