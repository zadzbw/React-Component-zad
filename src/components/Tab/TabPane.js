/**
 * Created by zad on 17/2/6.
 */
import React, {PropTypes, Component} from 'react';

export default class TabPane extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div></div>
    );
  }
}
