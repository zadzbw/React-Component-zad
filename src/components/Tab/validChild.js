/**
 * Created by zad on 17/3/22.
 */
import React, {Component, Children, cloneElement} from 'react';

export default function (WrappedComponent, childName) {
  return class Valid extends Component {
    static displayName = `Valid(${WrappedComponent.name})`;

    render() {
      const {children, ...props} = this.props;
      const validChildren = Children.map(children, (child, i) => {
        if (child.type !== childName) {
          return null;
        }
        return cloneElement(child, {
          _key: child.key ? child.key : `${WrappedComponent.name}-${i}`,
        });
      });
      return (
        <WrappedComponent
          {...props}
          children={validChildren}
        />
      );
    }
  };
}
