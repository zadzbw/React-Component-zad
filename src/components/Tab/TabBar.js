/**
 * Created by zad on 17/3/20.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import {tabPrefix} from './Tab';

export default class TabBar extends Component {
  constructor(props) {
    super(props);
    this._tabChange = this::this._tabChange;
  }

  _tabChange() {
    const {_tabChange, _key} = this.props;
    _tabChange(_key);
  }

  render() {
    const {active, name} = this.props;
    const barClass = classNames(`${tabPrefix}-bar`, {
      [`${tabPrefix}-bar-active`]: active,
    });

    return (
      <div className={barClass} onClick={this._tabChange}>
        {name}
      </div>
    );
  }
}
