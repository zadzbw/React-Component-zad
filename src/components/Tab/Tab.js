/**
 * Created by zad on 17/2/6.
 */
import React, {PropTypes, Component, Children} from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import _findIndex from 'lodash/findIndex';
import isOneOf from '../../utils/isOneOf';
import validChild from './validChild';
import keyCode from './keyCode';
import TabItem from './TabItem';
import TabBar from './TabBar';

const _type = {
  type: ['card', 'inline'],
};
export const tabPrefix = 'zad-tab';

class Tab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      current: 'current' in this.props ? this.props.current : this.props.defaultCurrent,
    };
    this.adjustCurrent = this::this.adjustCurrent;
    this._tabChange = this::this._tabChange;
    this.getInfo = this::this.getInfo;
    this.setStyle = this::this.setStyle;
  }

  static propTypes = {
    current: PropTypes.string,
    defaultCurrent: PropTypes.string,
    onTabChange: PropTypes.func,
    type: PropTypes.oneOf(['card', 'inline']),
  };

  static defaultProps = {
    defaultCurrent: '1',
    onTabChange: (current) => undefined,
    type: 'card',
  };

  componentWillReceiveProps(nextProps) {
    if ('current' in nextProps) {
      this.setState({
        current: nextProps.current,
      });
    }
  }

  componentDidMount() {
    this.adjustCurrent();
    this.setStyle(true);
  }

  componentDidUpdate() {
    this.setStyle(false);
  }

  // 排除非法的current
  adjustCurrent() {
    const props = this.props;
    if ('defaultCurrent' in props) {
      const {defaultCurrent} = props;
      const keys = this.getInfo().keys;
      if (!isOneOf(keys, defaultCurrent)) {
        this.setState({
          current: keys[0],
        });
      }
    }
  }

  _tabChange(key) {
    const {current} = this.state;
    const isChange = current !== key;
    if (isChange) {
      if (!('current' in this.props)) {
        this.setState({
          current: key,
        });
      }
      this.props.onTabChange(key);
    }
  }

  getInfo() {
    const {children} = this.props;
    const {current} = this.state;
    const keys = Children.map(children, child => child.props._key);
    const index = _findIndex(keys, (key) => key === current);
    return {
      index,
      keys,
    };
  }

  // todo 想一个不操作dom的方法去改变style?
  setStyle(isFirst) {
    const activeIndex = this.getInfo().index;
    const tabBarElm = ReactDOM.findDOMNode(this[`TabBar-${activeIndex}`]);
    if (!tabBarElm) {
      return;
    }
    const tabContent = this.tabContent;
    const tabBarRect = tabBarElm.getBoundingClientRect();
    const barsWrapRect = this.barsWrap.getBoundingClientRect();
    const contentRect = tabContent.getBoundingClientRect();

    const bottomStyle = {
      width: tabBarRect.width - 2, // 1px border on both sides
      offset: tabBarRect.left - barsWrapRect.left + 1, // 1px border on left side
    };
    this.setBottomStyle(bottomStyle, isFirst);

    const contentStyle = {
      offset: -(contentRect.width * activeIndex),
    };
    this.setContentStyle(contentStyle, isFirst);
  }

  setBottomStyle = (style, isFirst) => {
    const barsBottom = this.barsBottom;
    barsBottom.style.width = `${style.width}px`;
    barsBottom.style.transform = barsBottom.style.WebkitTransform = `translateX(${style.offset}px)`;
    !isFirst && (barsBottom.style.transition = 'transform 0.3s cubic-bezier(0.645, 0.045, 0.355, 1)');
  };

  setContentStyle = (style, isFirst) => {
    const tabContent = this.tabContent;
    tabContent.style.marginLeft = `${style.offset}px`;
    !isFirst && (tabContent.style.transition = 'margin-left 0.3s cubic-bezier(0.645, 0.045, 0.355, 1)');
  };

  _keyDown = (e) => {
    const code = e.keyCode;
    if (isOneOf([keyCode.LEFT, keyCode.UP], code)) {
      e.preventDefault();
      const prevKey = this.getNextKey(false);
      this._tabChange(prevKey);
    } else if (isOneOf([keyCode.RIGHT, keyCode.DOWN], code)) {
      e.preventDefault();
      const nextKey = this.getNextKey(true);
      this._tabChange(nextKey);
    }
  };

  // 获取下一个key
  getNextKey = (flag = true) => {
    const index = this.getInfo().index;
    const keys = this.getInfo().keys;
    const len = keys.length;
    let key;
    if (flag) {
      key = keys[(index + 1) % len];
    } else {
      key = keys[(len + index - 1) % len];
    }
    return key;
  };

  render() {
    const {type, className, children} = this.props;
    const activeIndex = this.getInfo().index;

    const tabBars = Children.map(children, (child, i) => {
      return React.createElement(TabBar, {
        ...child.props,
        active: activeIndex === i,
        _tabChange: this._tabChange,
        ref: (a) => this[`TabBar-${i}`] = a,
      });
    });
    const tabItems = Children.map(children, (child, i) => {
      return React.cloneElement(child, {
        active: activeIndex === i,
      });
    });

    const tabClass = classNames(tabPrefix, {
      [`${tabPrefix}-${type}`]: type && isOneOf(_type.type, type)
    }, className);

    return (
      <div className={tabClass}>
        <div className={`${tabPrefix}-bars-list`} tabIndex="0" onKeyDown={this._keyDown}>
          <div className={`${tabPrefix}-bars-container`}>
            <div className={`${tabPrefix}-bars-wrap`} ref={(a) => this.barsWrap = a}>
              <div className={`${tabPrefix}-bars-bottom`} ref={(a) => this.barsBottom = a}/>
              {tabBars}
            </div>
          </div>
        </div>
        <div className={`${tabPrefix}-content`} ref={(a) => this.tabContent = a}>
          {tabItems}
        </div>
      </div>
    );
  }
}

export default validChild(Tab, TabItem);
