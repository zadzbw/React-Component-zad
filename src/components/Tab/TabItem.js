/**
 * Created by zad on 17/2/6.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import {tabPrefix} from './Tab';

const TabItem = ({active, children}) => {
  const itemClass = classNames(`${tabPrefix}-item`, {
    [`${tabPrefix}-item-active`]: active,
    [`${tabPrefix}-item-inactive`]: !active,
  });
  return (
    <div className={itemClass}>{children}</div>
  );
};

TabItem.propTypes = {
  active: PropTypes.bool,
};

export default TabItem;
