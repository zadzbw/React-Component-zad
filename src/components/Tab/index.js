/**
 * Created by zad on 17/2/6.
 */
import Tab from './Tab';
import TabItem from './TabItem';
import './Tab.less';

Tab.TabItem = TabItem;
export default Tab;
