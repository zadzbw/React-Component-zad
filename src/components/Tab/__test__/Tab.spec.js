/**
 * Created by zad on 17/3/23.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Tab from '../../Tab';
import TabItem from '../../Tab/TabItem';
import TabBar from '../../Tab/TabBar';
import {describe, it} from 'eslint/lib/testers/event-generator-tester';

// 不知为何高阶组件不能simulate
describe('Tab', () => {
  it('normal', () => {
    const wrapper = shallow(
      <Tab defaultCurrent={'33'}>
        <TabItem name="Tab 1111" key="11">11111</TabItem>
        <TabItem name="Tab 22" key="22">22222</TabItem>
        <div>12345</div>
        <TabItem name="Tab 333" key="33">33333</TabItem>
      </Tab>
    );

    expect(wrapper.shallow().hasClass('zad-tab-card')).toBeTruthy();

    expect(wrapper.shallow().find(TabBar)).toHaveLength(3);
    expect(wrapper.shallow().find(TabBar).at(2).shallow().hasClass('zad-tab-bar-active')).toBeTruthy();
    expect(wrapper.shallow().find(TabItem)).toHaveLength(3);
    expect(wrapper.shallow().find(TabItem).at(2).shallow().hasClass('zad-tab-item-active')).toBeTruthy();

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('inline', () => {
    const wrapper = shallow(
      <Tab defaultCurrent={'22'} type={'inline'}>
        <TabItem name="Tab 1111" key="11">11111</TabItem>
        <TabItem name="Tab 22" key="22">22222</TabItem>
        <div>12345</div>
        <TabItem name="Tab 333" key="33">33333</TabItem>
        <TabItem name="Tab 444" key="4">4444</TabItem>
      </Tab>
    );

    expect(wrapper.shallow().hasClass('zad-tab-inline')).toBeTruthy();

    expect(wrapper.shallow().find(TabBar)).toHaveLength(4);
    expect(wrapper.shallow().find(TabBar).at(1).shallow().hasClass('zad-tab-bar-active')).toBeTruthy();
    expect(wrapper.shallow().find(TabItem)).toHaveLength(4);
    expect(wrapper.shallow().find(TabItem).at(1).shallow().hasClass('zad-tab-item-active')).toBeTruthy();

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  })
});
