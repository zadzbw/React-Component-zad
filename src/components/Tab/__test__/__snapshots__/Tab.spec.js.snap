exports[`Tab inline 1`] = `
<div
  className="zad-tab zad-tab-inline">
  <div
    className="zad-tab-bars-list"
    onKeyDown={[Function]}
    tabIndex="0">
    <div
      className="zad-tab-bars-container">
      <div
        className="zad-tab-bars-wrap">
        <div
          className="zad-tab-bars-bottom"
          style={Object {}} />
        <TabBar
          _key="11"
          _tabChange={[Function]}
          active={false}
          name="Tab 1111" />
        <TabBar
          _key="22"
          _tabChange={[Function]}
          active={true}
          name="Tab 22" />
        <TabBar
          _key="33"
          _tabChange={[Function]}
          active={false}
          name="Tab 333" />
        <TabBar
          _key="4"
          _tabChange={[Function]}
          active={false}
          name="Tab 444" />
      </div>
    </div>
  </div>
  <div
    className="zad-tab-content"
    style={Object {}}>
    <TabItem
      _key="11"
      active={false}
      animate={true}
      name="Tab 1111">
      11111
    </TabItem>
    <TabItem
      _key="22"
      active={true}
      animate={true}
      name="Tab 22">
      22222
    </TabItem>
    <TabItem
      _key="33"
      active={false}
      animate={true}
      name="Tab 333">
      33333
    </TabItem>
    <TabItem
      _key="4"
      active={false}
      animate={true}
      name="Tab 444">
      4444
    </TabItem>
  </div>
</div>
`;

exports[`Tab normal 1`] = `
<div
  className="zad-tab zad-tab-card">
  <div
    className="zad-tab-bars-list"
    onKeyDown={[Function]}
    tabIndex="0">
    <div
      className="zad-tab-bars-container">
      <div
        className="zad-tab-bars-wrap">
        <TabBar
          _key="11"
          _tabChange={[Function]}
          active={false}
          name="Tab 1111" />
        <TabBar
          _key="22"
          _tabChange={[Function]}
          active={false}
          name="Tab 22" />
        <TabBar
          _key="33"
          _tabChange={[Function]}
          active={true}
          name="Tab 333" />
      </div>
    </div>
  </div>
  <div
    className="zad-tab-content"
    style={Object {}}>
    <TabItem
      _key="11"
      active={false}
      animate={true}
      name="Tab 1111">
      11111
    </TabItem>
    <TabItem
      _key="22"
      active={false}
      animate={true}
      name="Tab 22">
      22222
    </TabItem>
    <TabItem
      _key="33"
      active={true}
      animate={true}
      name="Tab 333">
      33333
    </TabItem>
  </div>
</div>
`;
